### PR TITLE
[lld][WebAssembly] Fix stub library deps causing LTO archive members to be required post-LTO

### LINKED ISSUE #16836: Link failing with attempt to add bitcode file after LTO
Happens with 2.0.34 onward, latest I tried is 3.1.9, haven't tried earlier ones

When we compile our project with `-flto` we get:
```
[build] wasm-ld: error: D:\SDK\emsdk\upstream\emscripten\cache\sysroot\lib\wasm32-emscripten\lto\libc.a(fputc.o): attempt to add bitcode file after LTO.
```

I've looked at the relevant LLVM change log here https://reviews.llvm.org/rGb062fe181625bd1944ca9ca2a58246ffd7cd3536
And the previous issues #11545 and #15506

We don't use `fputc`, so it's most likely being generated by this transform https://github.com/llvm/llvm-project/blob/35cc2ec4ed4a7a0862e62ad009d561f61e2d12c3/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp#L2887

`fwrite(S,1,1,F) -> fputc(S[0],F)`

Adding a reference to it with `Wl,-u,fputc` fixes the issue, but I thought this has been fixed since #11545 is closed. 

Does the non-LTO symbol list in `system_libs.py` need updating maybe to keep up with any transforms that may emit these symbols late? Or is there another fix for this on the horizon?

### PR DESCRIPTION
**Pull request by sbc100**
Fixes: https://github.com/emscripten-core/emscripten/issues/16836

### CODE REVIEW

File: lld/test/wasm/lto/stub-library.s [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+# RUN: llvm-ar rcs %t/libfoo.a %t/foo.o
+# RUN: wasm-ld %t.o %t/libfoo.a %p/Inputs/stub.so -o %t.wasm
 # RUN: obj2yaml %t.wasm | FileCheck %s
 
-# The function `bar` is declared in stub.so and depends on `foo`, which happens
-# be in an LTO object.
-# This verifies that stub library dependencies (required exports) can be defined
-# in LTO objects.
+## The function `bar` is declared in stub.so and depends on `foo`, which happens
+## be in an LTO object, in an archive file.
```
Comment (kripken) [inline lld/test/wasm/lto/stub-library.s:10]:
```suggestion
## to be in an LTO object, in an archive file.
```

File: lld/test/wasm/lto/stub-library.s [hunk 2]
```diff
@@ -1,12 +1,15 @@
 # RUN: llvm-mc -filetype=obj -triple=wasm32-unknown-unknown -o %t.o %s
-# RUN: llvm-as %S/Inputs/foo.ll -o %t1.o
-# RUN: wasm-ld %t.o %t1.o %p/Inputs/stub.so -o %t.wasm
+# RUN: mkdir -p %t
+# RUN: llvm-as %S/Inputs/foo.ll -o %t/foo.o
+# RUN: rm -f %t/libfoo.a
+# RUN: llvm-ar rcs %t/libfoo.a %t/foo.o
+# RUN: wasm-ld %t.o %t/libfoo.a %p/Inputs/stub.so -o %t.wasm
```
Comment (kripken) [inline lld/test/wasm/lto/stub-library.s:6]: Do we not need the old test as well, to check the case of a raw object file and not one inside an archive?

File: lld/wasm/Driver.cpp [hunk 3]
```diff
@@ -949,6 +949,17 @@ static void processStubLibrariesPreLTO() {
           auto* needed = symtab->find(dep);
           if (needed ) {
             needed->isUsedInRegularObj = true;
+            // Like with handleLibcall we have extract any LTO archive members
```
Comment (kripken) [inline lld/wasm/Driver.cpp:952]:
```suggestion
            // Like with handleLibcall we have to extract any LTO archive members
```

File: lld/wasm/Driver.cpp [hunk 4]
```diff
@@ -949,6 +949,17 @@ static void processStubLibrariesPreLTO() {
...
+            // that might need to be exported due to stub library symbol being
```
Comment (kripken) [inline lld/wasm/Driver.cpp:953]:
```suggestion
            // that might need to be exported due to stub library symbols being
```

File: lld/wasm/Driver.cpp [hunk 5]
```diff
@@ -949,6 +949,17 @@ static void processStubLibrariesPreLTO() {
...
+            // used.  Without this the LTO object could be extracted during
+            // processStubLibraries, which is too late since LTO has already
+            // beeing performed at that point.
```
Comment (kripken) [inline lld/wasm/Driver.cpp:956]:
```suggestion
            // being performed at that point.
```

File: lld/wasm/Driver.cpp [hunk 6]
```diff
@@ -949,6 +949,17 @@ static void processStubLibrariesPreLTO() {
...
+            if (needed->isLazy() && isa<BitcodeFile>(needed->getFile())) {
+              if (!config->whyExtract.empty())
+                ctx.whyExtractRecords.emplace_back(toString(stub_file),
+                                                   needed->getFile(), *needed);
+              cast<LazySymbol>(needed)->extract();
+            }
```
Comment (kripken) [inline lld/wasm/Driver.cpp:962]: Perhaps this code could be shared with `handleLibcall`?

**Review by kripken — APPROVED** [review]

File: lld/test/wasm/lto/stub-library.s (hunk 2 above)
Comment (sbc100) [reply lld/test/wasm/lto/stub-library.s]: Done

**Review by kripken — APPROVED** [review]

**Comment by dschuff** [comment]
This seems to have caused a regression in the LTO versions of emscripten's [test_lazy_load_code_unconditional](https://github.com/emscripten-core/emscripten/blob/main/test/test_core.py#L8363) test (see [here](https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8740294891369218289/+/u/Emscripten_testsuite__LTO_/stdout) for a build log).
Before the change the base size is 4419 and the lazy size is 9140, but after they go up to 9162 and 13019 (with the test failing because the first is more than 60% of the lazy size). 
I'm not really familiar with this test, but it seems to be checking that the optimizer can eliminate a lot of the size of the first module (which I guess it does, but not 60%). But either way it seems like a large size regression

**Comment by kripken** [comment]
I confirmed the code size regression mentioned in the last comment has been resolved (almost entirely, at least: current sizes are 5174 and 9786, but that's very close to before the regression).

**Pull request merged**